### PR TITLE
CLI and Launcher documentation

### PR DESCRIPTION
**Pull request by cescoffier**
Provide the CLI and Launcher documentation

### CODE REVIEW

**Comment by cescoffier** [comment]
@vietj you asked for it... please review.

File: src/main/asciidoc/java/index.adoc [hunk 1]
```diff
@@ -932,13 +935,13 @@ java -jar my-application-fat.jar
 There is nothing really Vert.x specific about this, you could do this with any Java application
 
 You can either create your own main class and specify that in the manifest, but it's recommended that you write your
-code as verticles and use the Vert.x `Starter` class as your main class. This is the same main class used when running
-Vert.x at the command line and therefore allows you to specify command line arguments, such as `-instances` in order
-to scale your application more easily.
+code as verticles and use the Vert.x `Launcher` class as your main class. This is the same main class used when
```
Comment (vietj) [inline src/main/asciidoc/java/index.adoc]: recommend : use {@link io...Lauchner} so user can go to Javadoc

File: src/main/asciidoc/java/index.adoc [hunk 2]
```diff
@@ -951,14 +954,66 @@ java -jar my-verticle-fat.jar -cluster -conf myconf.json -cp path/to/dir/conf/cl
...
 ----
-vertx -version
+vertx version
+----
+
+=== Other commands
+
+The `vertx` command line and the `Launcher` also provide other _commands_ in addition to `run` and `version`:
+
+You can create a `bare instance using:
```
Comment (vietj) [inline src/main/asciidoc/java/index.adoc]: missing ending back quote after bare.

File: src/main/asciidoc/java/index.adoc [hunk 3]
```diff
@@ -951,14 +954,66 @@ java -jar my-verticle-fat.jar -cluster -conf myconf.json -cp path/to/dir/conf/cl
...
+----
+
+You can also start an application in background using:
+
+[source]
+----
+java -jar my-verticle-fat.jar start -Dvertx-id=my-app-name
+----
+
+If `my-app-name` is not set, a random id will be used (and display on the command prompt). You can pass `run`
```
Comment (vietj) [inline src/main/asciidoc/java/index.adoc]: and display_ed_

File: src/main/asciidoc/java/index.adoc [hunk 4]
```diff
@@ -951,14 +954,66 @@ java -jar my-verticle-fat.jar -cluster -conf myconf.json -cp path/to/dir/conf/cl
...
+options to the `start` commands:
```
Comment (vietj) [inline src/main/asciidoc/java/index.adoc]: no trailing s in command_s_

File: src/main/asciidoc/java/index.adoc [hunk 5]
```diff
@@ -951,14 +954,66 @@ java -jar my-verticle-fat.jar -cluster -conf myconf.json -cp path/to/dir/conf/cl
...
+
+[source]
+----
+java -jar my-verticle-fat.jar list
 ----
 
+The `start`, `stop` and `list` command are also available from the `vertx` tool. The start` command supports a couple of options:
+
+ * `vertx-id` : the application id, uses a random UUID if not set
+ * `java-opts` : the Java Virtual Machine options, uses the `JAVA_OPTS` environment variable if not set.
```
Comment (vietj) [inline src/main/asciidoc/java/index.adoc]: perhaps say how options should be encoded for proper char escape ?

File: src/main/asciidoc/java/cli.adoc [hunk 6]
```diff
@@ -0,0 +1,206 @@
+Vert.x Core provides an API for parsing command line options passed to programs.
```
Comment (vietj) [inline src/main/asciidoc/java/cli.adoc]: rather say arguments than options

File: src/main/asciidoc/java/cli.adoc [hunk 7]
```diff
@@ -0,0 +1,206 @@
...
+
+It's also able to print help
+messages detailing the options available for a command line tool. Even if such features are far from
+the vert.x core topics, this API is used in the `link:../../apidocs/io/vertx/core/Launcher.html[Launcher]` class that you can use in _fat-jar_
```
Comment (vietj) [inline src/main/asciidoc/java/cli.adoc]: Vert.x

File: src/main/asciidoc/java/cli.adoc [hunk 8]
```diff
@@ -0,0 +1,206 @@
...
+and in the `vertx` command line tools. In addition, it's polyglot (can be used from any supported language) and is
+used in vert.x shell.
```
Comment (vietj) [inline src/main/asciidoc/java/cli.adoc]: Vert.x Shell

File: src/main/asciidoc/java/cli.adoc [hunk 9]
```diff
@@ -0,0 +1,206 @@
...
+* GNU like long options (ie. `du --human-readable --max-depth=1`)
+* Java like properties (ie. `java -Djava.awt.headless=true -Djava.net.useSystemProxies=true Foo`)
+* Short options with value attached (ie. `gcc -O2 foo.c`)
+* Long options with single hyphen (ie. `ant -projecthelp`)
+
+Using the CLI api is a 3-steps process:
+
+1. The definition of the command line interface
+2. The parsing of the user command line
+3. The interrogation
```
Comment (vietj) [inline src/main/asciidoc/java/cli.adoc]: maybe replace interrogation by query ?

File: src/main/asciidoc/java/cli.adoc [hunk 10]
```diff
@@ -0,0 +1,206 @@
...
+----
+CLI cli = CLI.create("some-name")
+    .addOption(new Option()
+        .setLongName("optional")
+        .setDefaultValue("hello")
+        .setDescription("an optional option with a default value"));
+----
+
+An option can be _hidden_ using the `link:../../apidocs/io/vertx/core/cli/Option.html#setHidden-boolean-[setHidden]` method. Hidden option are
+not listed in the usage.
```
Comment (vietj) [inline src/main/asciidoc/java/cli.adoc]: yet can still be passed on the command line

File: src/main/asciidoc/java/cli.adoc [hunk 11]
```diff
@@ -0,0 +1,206 @@
...
+[source,java]
+----
+CommandLine commandLine = cli.parse(userCommandLineArguments);
+----
+
+The `link:../../apidocs/io/vertx/core/cli/CLI.html#parse-java.util.List-[parse]` method returns a `link:../../apidocs/io/vertx/core/cli/CommandLine.html[CommandLine]`
+object containing the values. By default, it validates the user command line and checks that each mandatory options
+and arguments have been set as well as the number of values received by each option. You can disable the
+validation by passing `false` as second parameter of `link:../../apidocs/io/vertx/core/cli/CLI.html#parse-java.util.List-boolean-[parse]`.
+
```
Comment (vietj) [inline src/main/asciidoc/java/cli.adoc]: This is useful if you want to check an argument or option is present even if the parsed command line is invalid.

File: src/main/asciidoc/java/index.adoc [hunk 12]
```diff
@@ -1205,4 +1260,59 @@ Any service can have potentially vulnerabilities whether it's written using Vert
...
+To use the `link:../../apidocs/io/vertx/core/Launcher.html[Launcher]` class in a _fat-jar_ just set the `Main-Class` of the _MANIFEST_ to
+`io.vertx.core.Launcher`. In addition, set the `Main-Verticle` _MANIFEST_ entry to the name of your main verticle.
+
+By default, it executed the `run` command. However, you can configure the default command by setting the
+`Main-Command` _MANIFEST_ entry. The default command is used if the _fat jar_ is launched without a command.
+
+=== Sub-classing the Launcher
+
+You can also create a sub-class of `link:../../apidocs/io/vertx/core/Launcher.html[Launcher]` to start your application. The class has been
+designed to be easily extendable.
```
Comment (vietj) [inline src/main/asciidoc/java/index.adoc]: extensible

**Comment by cescoffier** [comment]
@vietj - I've updated the code to fix your comment. Can you merge ?

**Comment by vietj** [comment]
yes

**Pull request merged**